# Отчёт_08_02_2021

## Изменение главной директории (выполнять только один раз!)

In [1]:
import sys
import os
os.chdir('..//')
print(os.getcwd())

C:\Users\NightHegg\Desktop\projects\SRW


## Подключение необходимых библиотек

In [2]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import scipy as sp

from scr.class_basic_method import basic_method
from scr.class_schwarz_multiplicative import schwarz_multiplicative
from scr.class_schwarz_additive import schwarz_additive
from scr.class_schwarz_two_level_additive import schwarz_two_level_additive

## Функция для получения результатов

In [3]:
def calculate_iters(cur_task, cur_method, list_mesh, list_subds):
    dict_result = {}
    dict_result_sigma = {}
    for cur_mesh in list_mesh:
        dict_iters_temp = {}
        dict_sigma_temp = {}
        for cur_amnt_subds in list_subds:
            obj = cur_method(cur_task, cur_mesh, cur_amnt_subds, coef_convergence)
            obj.get_solution()
            
            array_amnt_subds = np.prod(np.array(cur_amnt_subds))
            styled_amnt_subds = f"{array_amnt_subds} области" if array_amnt_subds < 5 else f"{array_amnt_subds} областей"
            
            dict_iters_temp[styled_amnt_subds] = obj.amnt_iterations
            dict_sigma_temp[styled_amnt_subds] = obj.get_special_sigma()
            
        dict_result[f"{cur_mesh**2} точек"] = dict_iters_temp
        dict_result_sigma[f"{cur_mesh**2} точек"] = dict_sigma_temp
    return dict_result, dict_result_sigma


def calculate_conv(cur_task, cur_method, cur_mesh, cur_subds, list_coef):
    dict_result = {}
    dict_result_temp = {}
    for cur_coef in list_coef:
        obj = cur_method(cur_task, cur_mesh, cur_subds, cur_coef)
        obj.get_solution()
        dict_result_temp[f"{cur_coef:.0e}"] = obj.amnt_iterations
    dict_result["Количество итераций"] = dict_result_temp
    return dict_result

### Предоставленные задачи:

- закрепление с трёх сторон;
- закрепление с двух сторон;
<div>
<img src="img/tasks_Rodin.jpg" width="250"/>
</div>

## Первая задача:

In [4]:
cur_task = 1
coef_convergence = 1e-4

list_subds = [[2, 1], [4, 1], [8, 1]]
list_mesh = [20, 25, 30]

test_mesh = 25
test_subds = [4, 1]
list_coef = [1e-3, 1e-4, 1e-5]

### Мультипликативный метод Шварца:

In [5]:
dict_iters_sm, dict_sigma_sm = calculate_iters(cur_task, schwarz_multiplicative, list_mesh, list_subds)

Таблица количества итераций:

In [6]:
pd.DataFrame.from_dict(dict_iters_sm)

400 точек  625 точек  900 точек
2 области           9          9          8
4 области          25         25         26
8 областей        113         83        106

Таблица разницы для осевого напряжения:

In [7]:
pd.DataFrame.from_dict(dict_sigma_sm)

400 точек                       625 точек  \
2 области   min = 9.19e+02, max = 2.11e+02  min = 1.09e+03, max = 3.14e+02   
4 области   min = 7.37e+02, max = 1.97e+02  min = 6.70e+02, max = 1.37e+02   
8 областей  min = 5.61e+02, max = 2.77e+02  min = 6.12e+02, max = 1.65e+02   

                                 900 точек  
2 области   min = 9.96e+02, max = 2.11e+02  
4 области   min = 7.17e+02, max = 2.10e+02  
8 областей  min = 8.75e+02, max = 3.50e+02

#### Дополнительные задачи:

In [8]:
dict_conv_sm = calculate_conv(cur_task, schwarz_multiplicative, test_mesh, test_subds, list_coef)

Таблица зависимости количества итераций от критерия сходимости (625 точек, 4 подобласти):

In [9]:
pd.DataFrame(dict_conv_sm)

Количество итераций
1e-03                   17
1e-04                   25
1e-05                   32

### Аддитивный метод Шварца:

In [10]:
dict_iters_sa, dict_sigma_sa = calculate_iters(cur_task, schwarz_additive, list_mesh, list_subds)

Таблица количества итераций:

In [11]:
pd.DataFrame.from_dict(dict_iters_sa)

400 точек  625 точек  900 точек
2 области          19         19         17
4 области          38         36         41
8 областей        161        107        159

Таблица разницы для осевого напряжения:

In [12]:
pd.DataFrame.from_dict(dict_sigma_sa)

400 точек                       625 точек  \
2 области   min = 1.27e+03, max = 2.02e+03  min = 3.26e+02, max = 2.26e+03   
4 области   min = 6.18e+02, max = 9.29e+02  min = 3.66e+02, max = 1.01e+03   
8 областей  min = 5.86e+02, max = 1.39e+03  min = 5.91e+02, max = 1.09e+03   

                                 900 точек  
2 области   min = 4.15e+02, max = 2.21e+03  
4 области   min = 7.12e+02, max = 1.07e+03  
8 областей  min = 7.08e+02, max = 1.46e+03

#### Дополнительные задачи:

In [13]:
dict_conv_sa = calculate_conv(cur_task, schwarz_additive, test_mesh, test_subds, list_coef)

Таблица зависимости количества итераций от критерия сходимости (625 точек, 4 подобласти):

In [14]:
pd.DataFrame(dict_conv_sa)

Количество итераций
1e-03                   26
1e-04                   36
1e-05                   46

### Двухуровневый аддитивный метод Шварца:

In [15]:
dict_iters_tsa, dict_sigma_tsa = calculate_iters(cur_task, schwarz_two_level_additive, list_mesh, list_subds)

Таблица количества итераций:

In [16]:
pd.DataFrame.from_dict(dict_iters_tsa)

400 точек  625 точек  900 точек
2 области          18         18         16
4 области          37         36         41
8 областей        170        106        200

Таблица разницы для осевого напряжения:

In [17]:
pd.DataFrame.from_dict(dict_sigma_tsa)

400 точек                       625 точек  \
2 области   min = 1.16e+03, max = 2.01e+03  min = 5.48e+02, max = 2.64e+03   
4 области   min = 6.82e+02, max = 1.02e+03  min = 3.48e+02, max = 9.37e+02   
8 областей  min = 7.48e+02, max = 7.39e+02  min = 6.11e+02, max = 1.13e+03   

                                 900 точек  
2 области   min = 4.97e+02, max = 2.62e+03  
4 области   min = 6.65e+02, max = 9.96e+02  
8 областей  min = 1.06e+03, max = 1.14e+03

#### Дополнительные задачи:

In [18]:
dict_conv_tsa = calculate_conv(cur_task, schwarz_two_level_additive, test_mesh, test_subds, list_coef)

Таблица зависимости количества итераций от критерия сходимости (625 точек, 4 подобласти):

In [19]:
pd.DataFrame(dict_conv_tsa)

Количество итераций
1e-03                   26
1e-04                   36
1e-05                   46